File: benchmark/nodejs/fannkuch_redux.js
function fannkuch(n) {
    let p = new Int32Array(n);
    let q = new Int32Array(n);
    let s = new Int32Array(n);
    let sign = 1;
    let maxflips = 0;
    let sumflips = 0;

    for (let i = 0; i < n; i++) {
        p[i] = i;
        q[i] = i;
        s[i] = i;
    }

    while (true) {
        let q1 = p[0];
        if (q1 !== 0) {
            for (let i = 1; i < n; i++) q[i] = p[i];
            let flips = 1;
            while (true) {
                let qq = q[q1];
                if (qq === 0) break;
                q[q1] = q1;
                if (q1 >= 3) {
                    let i = 1, j = q1 - 1;
                    while (i < j) {
                        let t = q[i];
                        q[i] = q[j];
                        q[j] = t;
                        i++;
                        j--;
                    }
                }
                q1 = qq;
                flips++;
            }
            if (flips > maxflips) maxflips = flips;
            sumflips += sign * flips;
        }

        if (sign === 1) {
            let t = p[1];
            p[1] = p[0];
            p[0] = t;
            sign = -1;
        } else {
            let t = p[1];
            p[1] = p[2];
            p[2] = t;
            sign = 1;
            let k = 2;
            for (; k < n; k++) {
                s[k]--;
                if (s[k] !== 0) break;
                s[k] = k;
                let t0 = p[0];
                for (let m = 0; m <= k; m++) {
                    p[m] = m < k ? p[m + 1] : t0;
                }
            }
            if (k === n) break;
        }
    }
    return `${sumflips}_${maxflips}`;
}

const start = Date.now();
const res = fannkuch(9);
const end = Date.now();

console.log(`Result: ${res}`);
console.log(`Time: ${end - start}ms`);

File: benchmark/src/python/fannkuch_redux.py
def fannkuch(n):
    p = list(range(n))
    q = list(range(n))
    s = list(range(n))
    sign = 1
    maxflips = 0
    sumflips = 0

    while True:
        q1 = p[0]
        if q1 != 0:
            for i in range(1, n):
                q[i] = p[i]
            flips = 1
            while True:
                qq = q[q1]
                if qq == 0:
                    break
                q[q1] = q1
                if q1 >= 3:
                    i, j = 1, q1 - 1
                    while i < j:
                        q[i], q[j] = q[j], q[i]
                        i += 1
                        j -= 1
                q1 = qq
                flips += 1
            if flips > maxflips:
                maxflips = flips
            sumflips += sign * flips
            
        if sign == 1:
            p[1], p[0] = p[0], p[1]
            sign = -1
        else:
            p[1], p[2] = p[2], p[1]
            sign = 1
            for i in range(2, n):
                s[i] -= 1
                if s[i] != 0:
                    break
                s[i] = i
                t = p[0]
                for j in range(i + 1):
                    p[j] = p[j + 1] if j < i else t
            else:
                break
    return sumflips, maxflips

import time
start = time.time()
sf, mf = fannkuch(9)
end = time.time()
print(f"Result: {sf}_{mf}")
print(f"Time: {int((end - start)*1000)}ms")
